OpenRA.Mods.Common/Projectiles/GravityBomb.cs

```diff
@@ -94,10 +98,25 @@ public void Tick(World world)
 			lastPos = pos;
 			pos += velocity;
 			velocity += acceleration;
+			var distanceAboveTerrain = world.Map.DistanceAboveTerrain(pos);
+			var shouldExplode = false;
+
+			// Check we're above the ground first
+			// We don't want to accidentally go through the floor to hit somthing
+			if (distanceAboveTerrain <= WDist.Zero)
+			{
+				pos += new WVec(0, 0, -distanceAboveTerrain.Length);
+				shouldExplode = true;
+			}
+
+			if (BlocksProjectiles.AnyBlockingActorsBetween(world, lastPos, pos, info.Width, out var blockedPos))
```

Comment:
I haven't looked at `AnyBlockingActorsBetween` in a while, but since it only asks for projectile _width_ I wouldn't be surprised if the check doesn't work as well on the vertical axis, if at all, since it was written mostly with ground combat in mind. That's something we need to check and potentially fix before we merge this.

Reply:
I suspect you're right - I just had a look into the guts of AnyBlockingActorsBetween, and it uses FindActorsOnLine - which has the Z values pinned at 0.

I'd be worried about the level of regression it'd cause to take the issue deeper into there, so is it worth having gravity bombs just implement their own logic for if there are actors blocking their path, that also checks the Z axis?

Reply:
I decided to start trying to hack it together just to see how big the blast radius was on trying to add Z as a factor on the existing functionality. Most functions I kicked over assumed a single plane and only dealt with X and Y axis - made even better because a lot of them deal with quite knotty math, which means adding extra optional parameters and more math makes them even harder to parse.

I'll stash it for now and have a look at how a stripped-down implementation just for (and on) gravity bombs looks, with the ideal that if it works there it might be worth then implementing into shared cube.

Reply:
I've had a go at implementing this functionality. I've tried to pull shared calculations out to private methods where I can to avoid just copy pasting code around, whilst still creating some new methods so as not to cause issues with regression. 

There was one piece of shared code changed, in MinimumPointLineProjection - luckily for me the WPos.Lerp functionality already checks for the Z Axis

I performed some regression tests across RA, TS and TD, and all three appear to be working fine - I think I was most concerned with missiles, but didn't notice any issues.

I think if we can get this implemented it could be really handy for other projectiles in the future, especially since on my journey of discovery on this, it looks like a good chunk of the code expects everything to be on just one plane.

![image](https://user-images.githubusercontent.com/18598259/117578333-c429b880-b0e5-11eb-96ff-246fbccb8648.png)